api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,629 @@
+			if err != nil {
+				return nil, err
+			}
+
+			if !hasResourcesFor(consumer, newProvider) {
+				text := fmt.Sprintf("Resource '%s' cannot fit into an empty provider!", consumer.ID)
+				text += "\nThe instance size in your environment is too small to run this resource."
+				text += "\nPlease increase the instance size for your environment"
+				err := fmt.Errorf(text)
+				errs = append(errs, err)
+				continue
+			}
+
+			subtractResourcesFor(consumer, newProvider)
+			providers = append(providers, newProvider)
+		}
+	}
+
+	// check if we need to scale down
+	unusedProviders := []*models.ResourceProvider{}
+	for _, provider := range providers {
+		if !provider.InUse {
+			unusedProviders = append(unusedProviders, provider)
+		}
+	}
+
+	desiredScale := len(providers) - len(unusedProviders)
+
+	actualScale, err := e.scaleTo(clusterName, desiredScale, unusedProviders)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	info := &models.ScalerRunInfo{
+		EnvironmentID:           clusterName,
+		PendingResources:        resourceConsumerModels(consumers),
+		ResourceProviders:       resourceProviderModels(providers),
+		ScaleBeforeRun:          scaleBeforeRun,
+		DesiredScaleAfterRun:    desiredScale,
+		ActualScaleAfterRun:     actualScale,
+		UnusedResourceProviders: len(unusedProviders),
+	}
+
+	return info, errors.MultiError(errs)
+}
+
+func resourceConsumerModels(consumers []models.ResourceConsumer) []models.ResourceConsumer {
+	consumerModels := make([]models.ResourceConsumer, len(consumers))
+	for i, consumer := range consumers {
+		newConsumer := models.ResourceConsumer{
+			ID:     consumer.ID,
+			Memory: consumer.Memory,
+			Ports:  consumer.Ports,
+		}
+		consumerModels[i] = newConsumer
+	}
+
+	return consumerModels
+}
+
+func resourceProviderModels(providers []*models.ResourceProvider) []models.ResourceProvider {
+	providerModels := make([]models.ResourceProvider, len(providers))
+	for i, provider := range providers {
+		newProvider := models.ResourceProvider{
+			ID:              provider.ID,
+			InUse:           provider.InUse,
+			UsedPorts:       provider.UsedPorts,
+			AvailableMemory: provider.AvailableMemory,
+		}
+		providerModels[i] = newProvider
+	}
+
+	return providerModels
+}
+
+func (e *EnvironmentScaler) getResourceProviders(clusterName string) ([]*models.ResourceProvider, error) {
+	listContainerInstancesInput := &ecs.ListContainerInstancesInput{}
+	listContainerInstancesInput.SetCluster(clusterName)
+
+	containerInstanceARNs := []*string{}
+	listContainerInstancesPagesFN := func(output *ecs.ListContainerInstancesOutput, lastPage bool) bool {
+		containerInstanceARNs = append(containerInstanceARNs, output.ContainerInstanceArns...)
+
+		return !lastPage
+	}
+
+	if err := e.Client.ECS.ListContainerInstancesPages(listContainerInstancesInput, listContainerInstancesPagesFN); err != nil {
+		return nil, err
+	}
+
+	describeContainerInstancesInput := &ecs.DescribeContainerInstancesInput{}
```

Comment:
Might as well check for a 0 length of `containerInstanceARNs` returned. If 0 we don't need to proceed any further. I.e., 

```go
result := []*models.ResourceProvider{}
if len(containerInstanceARNs) == 0 {
	return result, nil
}
```